src/lib.rs

```diff
@@ -248,31 +258,307 @@ impl<T, const N: usize> ArrayVec<T, { N }> {
+    ///
+    /// assert_eq!(vector.force_insert(0, 42), None);
+    /// assert_eq!(vector.force_insert(0, 24), None);
+    /// assert_eq!(&vector, [24, 42].as_ref());
+    /// ```
+    ///
+    /// But when the vector's full, we get a different result:
+    ///
+    /// ```rust
+    /// use const_arrayvec::ArrayVec;
+    /// let mut vector = ArrayVec::from([
+    ///     "He".to_owned(),
+    ///     "ya".to_owned(),
+    /// ]);
+    ///
+    /// let out = vector
+    ///     .force_insert(1, "llo".to_owned())
+    ///     .unwrap();
+    ///
+    /// assert_eq!(&out, "ya");
+    /// assert_eq!(&vector, ["He".to_owned(), "llo".to_owned()].as_ref());
+    /// ```
+    pub fn force_insert(&mut self, index: usize, item: T) -> Option<T> {
+        let len = self.len();
+
+        // bound checks
+        if index > len {
+            out_of_bounds!("force_insert", index, len);
+        }
+
+        let result;
+
+        if self.is_full() {
+            // bound checks (here: N == len)
+            if index >= N {
+                out_of_bounds!("force_insert", index, len);
+            }
+
+            unsafe {
+                // Store the last item before it's removed.
+                result = Some(ptr::read(self.as_ptr().add(len - 1)));
+
+                // The last element should be removed so we shouldn't
+                // copy it.
+                self.insert_unchecked_keep_len(index, item, len - 1);
+            }
+        } else {
+            // bound checks
+            if index > len {
+                out_of_bounds!("force_insert", index, len);
+            }
+
+            // Since nothing's going to be removed, the vector's size
+            // is going to be increased and nothing will be returned.
+            unsafe {
+                self.insert_unchecked(index, item, len);
+                result = None;
+            }
+        }
+
+        result
+    }
+
+    /// Insert an item into the vector without checking if the index is
+    /// valid or if the vector isn't full or the vector's length.
+    ///
+    /// # Safety
+    ///
+    /// If you plan on using this function, you need to check for the
+    /// 3 previously mentioned conditions yourself before calling this
+    /// method.
+    #[inline]
+    pub unsafe fn insert_unchecked(
+        &mut self,
+        index: usize,
+        item: T,
+        len: usize,
+    ) {
+        self.insert_unchecked_keep_len(index, item, len);
+        self.set_len(len + 1);
+    }
+
+    /// Insert an item into the vector without checking if the index is
+    /// valid or if the vector isn't full or the vector's length and
+    /// without incrementing the vector's length.
+    ///
+    /// # Safety
+    ///
+    /// If you plan on using this function, you need to check for the
+    /// 3 previously mentioned conditions yourself before calling this
+    /// method. You also need to increment the vector's length afterward
+    /// yourself.
+    pub unsafe fn insert_unchecked_keep_len(
```

Comment:
Does this need to be `pub`? It feels like a footgun, and people should get by just fine using things like `insert()` or `insert_unchecked()`. Or if they really want to do `unsafe` things like this they can use `as_ptr()` and write to the backing buffer themselves.

Reply:
You're right, I'll make it private.